internal/cmd/marketplace/apply.go

```diff
@@ -0,0 +1,319 @@
+
+# Apply all the valid configuration files in the directory myFantasticGoTemplates to the Marketplace
+miactl marketplace apply -d myFantasticGoTemplates`
+)
+
+// ApplyCmd returns a new cobra command for adding or updating marketplace resources
+func ApplyCmd(options *clioptions.CLIOptions) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "apply { { -f file-path }... | { -d directory-path }... }",
+		Short:   "Create or update Marketplace items",
+		Long:    applyLong,
+		Example: applyExample,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if options.MarketplaceResourcesDirPath == "" && len(options.MarketplaceResourceFilePaths) == 0 {
+				return errors.New(`at least one of  "directory" or "file" must be set`)
+			}
+
+			restConfig, err := options.ToRESTConfig()
+			cobra.CheckErr(err)
+			client, err := client.APIClientForConfig(restConfig)
+			cobra.CheckErr(err)
+
+			var resourceFilesPaths []string
+			if len(options.MarketplaceResourceFilePaths) > 0 {
+				resourceFilesPaths = options.MarketplaceResourceFilePaths
+			}
+			if options.MarketplaceResourcesDirPath != "" {
+				resourceFilesPaths, err = buildPathsListFromDir(options.MarketplaceResourcesDirPath)
+				cobra.CheckErr(err)
+			}
+			applyReq, err := buildApplyRequest(resourceFilesPaths)
+			cobra.CheckErr(err)
+			outcome, err := applyMarketplaceResource(cmd.Context(), client, restConfig.CompanyID, applyReq)
+			cobra.CheckErr(err)
+
+			fmt.Println(buildOutcomeSummaryAsTables(outcome))
+
+			return nil
+		},
+	}
+
+	options.AddMarketplaceApplyFlags(cmd.Flags())
+	cmd.MarkFlagsMutuallyExclusive("file", "directory")
+
+	return cmd
+}
+
+const applyEndpoint = "/api/backend/marketplace/tenants/%s/resources"
+
+var (
+	errResWithoutName       = errors.New(`the required field "name" was not found in the resource`)
+	errNoValidFilesProvided = errors.New("no valid files were provided, see errors above")
+	errDuplicatedResName    = errors.New("some resources have duplicated name field")
+	errResNameNotAString    = errors.New(`the field "name" must be a string`)
+	errInvalidExtension     = errors.New("file has an invalid extension. Valid extensions are `.json`, `.yaml` and `.yml`")
+)
+
+func listFilesInDirRecursive(rootPath string) ([]string, error) {
+	var files []string
+	err := filepath.Walk(rootPath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() {
+			files = append(files, path)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return files, nil
+}
+
+func buildPathsListFromDir(dirPath string) ([]string, error) {
+	filesPaths, err := listFilesInDirRecursive(dirPath)
+	if err != nil {
+		return nil, err
+	}
+	filePaths := []string{}
+	for _, path := range filesPaths {
+		switch filepath.Ext(path) {
+		case encoding.YamlExtension:
+			fallthrough
+		case encoding.YmlExtension:
+			fallthrough
+		case encoding.JSONExtension:
```

Comment:
Way separate the three cases if they do the same thing? The `fallthrough` keyword is always a smell to me...

```suggestion
		case encoding.YamlExtension, encoding.YmlExtension, encoding.JSONExtension:
```